File: backend/src/main/java/com/gxs/bank/repository/UserRepository.java
package com.gxs.bank.repository;

import com.gxs.bank.model.User;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.stereotype.Repository;
import java.util.List;
import java.util.Optional;
import java.util.UUID;

@Repository
public interface UserRepository extends JpaRepository<User, UUID> {
    Optional<User> findByEmail(String email);
    boolean existsByEmail(String email);
    List<User> findByRole(User.Role role);
    long countByRole(User.Role role);
    List<User> findByKycStatus(User.KycStatus kycStatus);
    List<User> findByRoleAndKycStatus(User.Role role, User.KycStatus kycStatus);
    long countByKycStatus(User.KycStatus kycStatus);
}

File: backend-python/com/gxs/bank/repository/UserRepository.py
from sqlalchemy import func

from com.gxs.bank.model.User import User
from com.gxs.bank.repository._base import BaseRepository


class UserRepository(BaseRepository):
    def __init__(self, db):
        super().__init__(db, User)

    def findByEmail(self, email: str):
        return self.db.query(User).filter(User.email == email).first()

    def existsByEmail(self, email: str) -> bool:
        return self.findByEmail(email) is not None

    def findByRole(self, role):
        return self.db.query(User).filter(User.role == role).all()

    def countByRole(self, role):
        return int(self.db.query(func.count(User.id)).filter(User.role == role).scalar() or 0)

    def findByKycStatus(self, kycStatus):
        return self.db.query(User).filter(User.kycStatus == kycStatus).all()

    def findByRoleAndKycStatus(self, role, kycStatus):
        return self.db.query(User).filter(User.role == role, User.kycStatus == kycStatus).all()

    def countByKycStatus(self, kycStatus):
        return int(self.db.query(func.count(User.id)).filter(User.kycStatus == kycStatus).scalar() or 0)
